File: TEO/Mav.UI/Mav.Ui/src/app/services/i-statement.service.ts
import { Observable } from 'rxjs/Observable';
import { GenericResult } from '../models/generic.result.model';
import { UpdateStatementViewModel } from '../models/admin/update-statement.view.model';
import { StatementAnswerViewModel } from '../models/statement.answer.view.model';

export interface IStatementService {

    getStatementById<T>(id: number): Observable<T>;

    getStatementBySectionId<T>(id: number): Observable<T>;

    getStatementByAssessmentId<T>(id: number): Observable<T>;

    createStatement(newItem: UpdateStatementViewModel): Observable<GenericResult<number>>;

    updateStatement(item: UpdateStatementViewModel): Observable<GenericResult<number>>;

    deleteStatement(id: number): Observable<GenericResult<number>>;

    getUserStatementAnswers<T>(id: number, userid: string): Observable<T>;

    updateUserStatementAnswers(item: UpdateStatementViewModel): Observable<GenericResult<number>>;

}

File: TEO/Mav.Services/Abstract/IStatementService.cs
using Mav.Common.Models;
using Mav.Models.ViewModels;
using System;
using System.Collections.Generic;
using System.Text;
using System.Threading.Tasks;

namespace Mav.Services.Abstract
{
    public interface IStatementService
    {
        Task<T> GetStatementById<T>(int id);

        Task<T> GetStatementBySectionId<T>(int id);

        Task<T> GetStatementByAssessmentById<T>(int id);

        Task<GenericResult<T>> CreateStatement<T>(AddStatementViewModel m);

        Task<GenericResult<T>> UpdateStatement<T>(UpdateStatementViewModel m);

        Task<GenericResult<T>> DeleteStatement<T>(int id);

        Task<T> GetStatementAnswers<T>(int statementId, string userId);

        Task<GenericResult<T>> CreateStatementAnswer<T>(UserStatementAnswerViewModel m);

        Task<GenericResult<T>> UpdateStatementAnswer<T>(UserStatementAnswerViewModel m);
    }
}
